format/private/format.sh

```diff
@@ -123,160 +128,79 @@ case "$mode" in
-  echo "Formatting Jsonnet with jsonnetfmt..."
-  echo "$files" | tr \\n \\0 | xargs -0 $bin $jsonnetmode
-fi
-
-files=$(ls-files Java $@)
-bin=$(rlocation {{java-format}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting Java with java-format..."
-  # Setting JAVA_RUNFILES to work around https://github.com/bazelbuild/bazel/issues/12348
-  echo "$files" | tr \\n \\0 | JAVA_RUNFILES="${RUNFILES_MANIFEST_FILE%_manifest}" xargs -0 $bin $javamode
-fi
-
-files=$(ls-files Kotlin $@)
-bin=$(rlocation {{ktfmt}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting Kotlin with ktfmt..."
-  echo "$files" | tr \\n \\0 | xargs -0 $bin $ktmode
-fi
-
-files=$(ls-files Scala $@)
-bin=$(rlocation {{scalafmt}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting Scala with scalafmt..."
-  # Setting JAVA_RUNFILES to work around https://github.com/bazelbuild/bazel/issues/12348
-  echo "$files" | tr \\n \\0 | JAVA_RUNFILES="${RUNFILES_MANIFEST_FILE%_manifest}" xargs -0 $bin $scalamode
-fi
-
-files=$(ls-files Go $@)
-bin=$(rlocation {{gofmt}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting Go with gofmt..."
-  # gofmt doesn't produce non-zero exit code so we must check for non-empty output
-  # https://github.com/golang/go/issues/24230
-  if [ "$mode" == "check" ]; then
-    NEED_FMT=$(echo "$files" | tr \\n \\0 | xargs -0 $bin $gofmtmode)
-    if [ -n "$NEED_FMT" ]; then
-       echo "Go files not formatted:"
-       echo "$NEED_FMT"
-       exit 1
-    fi
-  else
-    echo "$files" | tr \\n \\0 | xargs -0 $bin $gofmtmode
+function run-lints {
+  local lang="$1" && shift
+  local fmtname="$1" && shift
+  local bin="$1" && shift
+  local args="$1" && shift
+  local tuser
+  local tsys
+
+  local files=$(ls-files "$lang" $@)
+  if [ -n "$files" ] && [ -n "$bin" ]; then
+    echo "Formatting ${lang} with ${fmtname}..."
+    case "$lang" in
+    'Protocol Buffer')
+        ( for file in $files; do
+          "$bin" $args $file >&2
+        done ; times ) | ( read _ _; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        ;;
+      Go)
+        # gofmt doesn't produce non-zero exit code so we must check for non-empty output
+        # https://github.com/golang/go/issues/24230
+        if [ "$mode" == "check" ]; then
+          GOFMT_OUT=$(mktemp)
+          (echo "$files" | tr \\n \\0 | xargs -0 "$bin" $args > "$GOFMT_OUT" ; times ) | ( read _ _; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+          NEED_FMT="$(cat $GOFMT_OUT)"
+          rm $GOFMT_OUT
+          if [ -n "$NEED_FMT" ]; then
+            echo "Go files not formatted:"
+            echo "$NEED_FMT"
+            exit 1
+          fi
+        else
+          ( echo "$files" | tr \\n \\0 | xargs -0 "$bin" $args >&2 ; times ) | ( read _ _; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        fi
+        ;;
+      Java)
+          # Setting JAVA_RUNFILES to work around https://github.com/bazelbuild/bazel/issues/12348
```

Comment:
nit: I wonder if it's possible to reduce the duplication, either by setting `JAVA_RUNFILES` in the default case, or setting it and then falling through to the default case

Reply:
i pulled some of the timing logic into another helper function, where possible. PTAL

Reply:
that's plenty of improvement, thanks!